doc/architecture/decisions/0013-signals-listeners-model-and-api.md

```diff
@@ -0,0 +1,55 @@
+# 13. Signal/Listener model and API
+
+Date: 2018-05-16
+
+## Status
+
+Proposed
+
+## Context
+
+In the go code we used polling from the UI to discover local state changes, be they in the source-chain or in the DHT.  We want to change this to an event based model where UI (or other clients, like bridged apps) can become listeners to signals.  Additionally there are system level signals (like warrants) as well as other app level definable signals that we want consumers of exposed holochain app APIs to be able to access on a push (i.e. listening) basis.
+
+Note that this is only about signals that can be listened to by container/composer context, i.e. by the client of the core_api, NOT by the app nodes themselves.
+
+## Decision
+
+We will extend the API in a way that's roughly equivalent to the [Signal-slot pattern](https://en.wikipedia.org/wiki/Signals_and_slots), i.e. in the following ways:
+
+1. In the DNA you can declare observable signals anywhere you can declare functions.  You can think of this almost identically to declaring a function except that it "goes the other way," i.e. a function def exposes an entry point where an signal def exposes an exit point.  It would look like this:
+
+``` javascript
+          "signals": [
+            {
+              "name": "Post",
+              "description: "signal emmited when a post is committed",
+              "params": null,
+              "sends": {
```

Comment:
Reading through this again, I wonder why there is ```params``` and ```sends```. I understood ```sends``` to be the arguments of the signal.. what is ```params``` then? Can we coalesce both to ```arguments```?

Reply:
Nope, actually sends is the equivalent of "returns" i.e. it's what the signal will send to you when triggered.  I meant for "params" to be what you as the listener can pass in when you register as a listener.  The idea was to create a space for parameterizing the listening, e.g. this would be a way to listen for Posts only from a specific person, and that parameter gets passed into the handler for the signal so it can do the filtering.

Perhaps a different name to be less confusing?

Reply:
Well, I don't fully understand your ```params``` then. Or I should rather say, I think to build what you want it gets too complex. I would suggest removing ```params``` and renaming ```sends``` to arguments, because for the listener/slot/handler it is what comes in as parameter / function argument.

Reply:
`params`  are like configuration variables. It is input information on how you want to receive the output (`sends`). For exemple when you subscribe to a newsletter you might be able to indicate the frequency with which you want ot receive the emails.

Reply:
Exactly what @ddd-mtl said.  Implementing this doesn't seem hard because all it means is that you have to provide the signaler the value passed in at registration time.  That doesn't seem hard.  Am I missing something think @lucksus?  

It is possible that this is a YAGNI, i.e. and I'm overgeneralizing, and we can implement it later if we want it.

Reply:
Hm, ok. What do you think about changing ```params``` to ```config-params``` to make this more clear? And ```sends``` -> ```arguments```? would make it clearer to me upon first read. I will go and approve this PR now nevertheless. don't want to block it any further.. ;)